cabal-install: check: Whitelist doc file extensions

wismill opened the pull request:
Fixes #8746

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

ulysses4ever reviewed (approved):
Awesome, thank you!

andreasabel commented on Cabal/src/Distribution/PackageDescription/Check.hs:
```diff
@@ -2545,19 +2545,21 @@ checkGlobFiles verbosity pkg root = do
-    isDesirableExtraDocFile paths path = map toLower basename `elem` paths
+    isDesirableExtraDocFile :: ([FilePath], [FilePath]) -> FilePath -> Bool
+    isDesirableExtraDocFile (basenames, extensions) path =
+      map toLower basename `elem` basenames && ext `elem` extensions
```
`ext` is not subjected to `map toLower` here.  I wonder how this plays out in case-insensitive filesystems (e.g. on Windows).
It seems that some folks call the file "CHANGELOG.TXT", e.g. https://github.com/moodle/moodle/blob/0780e87f06eb834786f562316ad9699480bcd24e/lib/tcpdf/CHANGELOG.TXT

wismill replied:
Fixed and added test.

It is worth to update hackage-server as well (see haskell/hackage-server#1179).

andreasabel reviewed (commented):
I think extensions should include `.TXT` and maybe more generally cater for case-insensitive file systems.

andreasabel reviewed (approved):
Thanks!

Mikolaj commented:
One more thing in the test seems to still require an update. Also please rebase to take advantage of the newest CI fix.

wismill commented:
Rebased and (hopefully) fixed.

File order does not seem to be deterministic or is platform-dependent. As we have two similar test cases, I fixed it by keeping only one file in each case.

Mikolaj commented:
Am I correct this needs a backport to 3.10?

wismill commented:
Yes _only_ if you backported #8657.

Mikolaj commented:
I did not. But it was merged before 3.10 has been cut. Wasn't it?
